pkg/vault/contracts/authorizer/TimelockAuthorizer.sol

```diff
@@ -776,10 +820,15 @@ contract TimelockAuthorizer is IAuthorizer, IAuthentication, ReentrancyGuard {
     ) external {
         InputHelpers.ensureInputLengthMatch(actionIds.length, where.length);
         for (uint256 i = 0; i < actionIds.length; i++) {
-            // For permissions that have a delay when granting, `canGrant` will return false. `scheduleGrantPermission`
-            // will succeed as it checks `isGranter` instead.
-            // Note that `canGrant` will return true for the executor if the permission has a delay.
-            require(canGrant(actionIds[i], msg.sender, where[i]), "SENDER_IS_NOT_GRANTER");
+            if (_grantDelays[actionIds[i]] == 0) {
+                require(isGranter(actionIds[i], msg.sender, where[i]), "SENDER_IS_NOT_GRANTER");
+            } else {
+                // Some actions may have delays associated with granting them - these permissions cannot be granted
+                // directly, even if the caller is a granter, and must instead be scheduled for future execution via
+                // `scheduleGrantPermission`.
+                require(msg.sender == address(_executor), "GRANT_MUST_BE_SCHEDULED");
+            }
```

Comment:
I found `canGrant` a bit confused because of the existence of `isGranter`, but the helper did make this for loop more readable.

How about keeping the helper, maybe thinking about a new name?
On the other hand, I thought it was a useful helper to expose.

Reply:
You mean `canGrant`? I think it's a terrible concept :/ We are forced to have `canPerform`, but I'd not have it unless we needed to. How did putting this behind a function make it more readable?

Reply:
I think the name was not great, but it was convenient at least in this context. The structure was straightforward: iterate action IDs, can they be granted? if yes, go, otherwise revert.
But yeah, doing it like this exposes two different error messages which is good.

Feel free to disregard the comment; perhaps I'd suggest moving the comment inside the `else` block above the `if`.
How about something like

```
  // If the current action does not have an associated delay, the permission can be granted right away provided that the caller is a granter.
  // On the other hand, some actions may have delays associated with granting them - these permissions cannot be granted
  // directly, even if the caller is a granter, and must instead be scheduled for future execution via
  // `scheduleGrantPermission`.
```

Reply:
The thing is 'can they be granted' really means 'can they be granted in the context of a non-scheduled action', which is a complex and weird concept - I find the code just being there makes everything much more obvious.